EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+---
+eip: 9999
+title: Block-level warming
+description: Warm addresses and storage keys over the duration of a block
+author: Toni Wahrstätter (@nerolation), Alex Stokes (@ralexstokes), Ansgar Dietrichs (@adietrichs)
+discussions-to: ""
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-01-15
+---
+
+## Abstract
+
+This proposal introduces block-level address and storage key warming, allowing accessed addresses and storage keys to maintain their "warm" status throughout an entire block's execution. Accessed slots can be effectively cached at the block level, allowing for this opimization.
+
+## Motivation
+
+Currently, the EVM's storage slot warming mechanism operates at the transaction level, requiring each transaction to "warm up" slots independently, even when accessing the same storage locations within the same block. This design does not take advantage of the fact that modern node implementations can effectively cache storage access patterns at the block level. By extending the slot warming duration to the block level, we can:
+
+1. Reduce redundant warming costs for frequently accessed slots
+2. Better align gas costs with actual computational overhead
+3. Improve overall network throughput without compromising security.
+
+As of Jan. 2025, the empirically messured efficiency gains are around 5%.
+
+
+## Specification
+
+### Mechanics
+
+When a storage slot is accessed within a block:
+1. The first access to a slot in a block incurs the cold access cost as of [EIP-2929](./eip-2929).
+2. All subsequent accesses to the same slot within the same block incur only the warm access cost as of [EIP-2929](./eip-2929).
+3. The warm/cold status resets at block boundaries
+
+### Block Processing
+
+1. At the start of each block:
+   * Initialize two empty sets `block_level_accessed_addresses` and `block_level_accessed_storage_keys`
```

Comment:
I know this is rather implicit, but the `block_level_accessed_storage_keys` is rather an item of a Map from addresses to the Set of accessed storage keys of that address.

Reply:
Yeah true, although, one could also concat contract+storage key and treat the resulting bytes as a unique storage key